src/data/pnat/factor_finsupps.lean

```diff
@@ -0,0 +1,542 @@
+/-
+Copyright (c) 2020 Aaron Anderson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Aaron Anderson.
+ -/
+
+import data.finsupp
+import data.pnat.factors
+import order.lattice
+import order.order_iso
+
+/-!
+# Factor Finsupps
+
+A new version of prime factorisation and some facts about lattice isomorphisms
+
+## Main Definitions
+- `factor_finsupp` sends a `pnat` to its prime factorization,
+  expressed as a function from the primes to their multiplicities.
+
+## Implementation Notes
+- `factor_finsupp` is based on `pnat.factor_multiset`.
+- Many facts about it are proved by constructing a lattice isomorphism between `multiset`s
+  and `finsupp`s to ℕ.
+
+-/
+
+set_option old_structure_cmd true
+
+open_locale classical
+open_locale big_operators
+
+
+noncomputable theory
+
+section basics
+
+instance finsupp.coe_primes_to_pnat : has_coe (nat.primes →₀ ℕ) (ℕ+ →₀ ℕ) :=
+⟨finsupp.map_domain coe⟩
+
+lemma prime_multiset.to_pnat_multiset_to_multiset (x : nat.primes →₀ ℕ) :
+  prime_multiset.to_pnat_multiset x.to_multiset = (x : ℕ+ →₀ ℕ).to_multiset :=
+finsupp.to_multiset_map x coe
+
+/-- Take the product of a function of prime powers with exponents given by a finsupp. -/
+def finsupp.primes_prod_apply_pow {α : Type} [comm_monoid α] (x : nat.primes →₀ ℕ) (f : ℕ+ → α) : α :=
+x.prod (λ p : nat.primes, λ k : ℕ, f (p ^ k))
+
+/-- Take the product of prime powers with exponents given by a finsupp. -/
+def finsupp.primes_prod_pow (x : nat.primes →₀ ℕ) := x.primes_prod_apply_pow id
+
+lemma finsupp.primes_prod_pow_eq (x : nat.primes →₀ ℕ) :
+  x.primes_prod_pow= x.prod (λ p : nat.primes, λ k : ℕ, (p ^ k)) := rfl
+
+lemma finsupp.primes_prod_pow_eq_coe_prod_pow (x : nat.primes →₀ ℕ) :
+  x.primes_prod_pow= (x : ℕ+ →₀ ℕ).prod pow :=
+begin
+  change x.primes_prod_pow = (finsupp.map_domain coe x).prod pow,
+  rw [finsupp.prod_map_domain_index, finsupp.primes_prod_pow_eq], simp,
+  intros, rw pow_add,
+end
+
+lemma finsupp.primes_prod_pow_eq_prod_to_multiset (x : nat.primes →₀ ℕ) :
+  x.primes_prod_pow= prime_multiset.prod x.to_multiset :=
+begin
+  rw [prime_multiset.prod, finsupp.primes_prod_pow_eq_coe_prod_pow,
+    ← finsupp.prod_to_multiset, ← prime_multiset.to_pnat_multiset_to_multiset], refl
+end
+
+/-- The value of this finsupp at a prime is the multiplicity of that prime in n. -/
+def factor_finsupp (n : ℕ+) : nat.primes →₀ ℕ := n.factor_multiset.to_finsupp
+
+@[simp]
+lemma factor_finsupp_to_multiset_eq_factor_multiset {n : ℕ+} :
+  (factor_finsupp n).to_multiset = n.factor_multiset :=
+by { unfold factor_finsupp, simp }
+
+end basics
+
+section finsupp_lattice
+-- Should this go to data/finsupp.lean? New file data/finsupp/order.lean?
```

Comment:
Could you perhaps make a separate PR that just does this stuff? I think I'd like to be able to consider these lattice isomorphisms without having to think about primes or natural numbers at all.

Reply:
Sure. I'm not sure if there's a most-elegant way to do this with git, though.

Reply:
#3335